src/backend/concurrency/eager_write_txn_manager.cpp

```diff
@@ -0,0 +1,672 @@
+
+        new_tile_group_header->SetTransactionId(new_version.offset,
+                                                INVALID_TXN_ID);
+        // AtomicSetOnlyTxnId(tile_group_header, tuple_slot, INITIAL_TXN_ID);
+        tile_group_header->SetTransactionId(tuple_slot, INITIAL_TXN_ID);
+
+      } else if (tuple_entry.second == RW_TYPE_DELETE) {
+        ItemPointer new_version =
+            tile_group_header->GetNextItemPointer(tuple_slot);
+
+        auto new_tile_group_header =
+            manager.GetTileGroup(new_version.block)->GetHeader();
+
+        new_tile_group_header->SetBeginCommitId(new_version.offset, MAX_CID);
+        new_tile_group_header->SetEndCommitId(new_version.offset, MAX_CID);
+
+        COMPILER_MEMORY_FENCE;
+
+        tile_group_header->SetEndCommitId(tuple_slot, MAX_CID);
+
+        // reset the item pointers.
+        tile_group_header->SetNextItemPointer(tuple_slot, INVALID_ITEMPOINTER);
+        new_tile_group_header->SetPrevItemPointer(new_version.offset, INVALID_ITEMPOINTER);
+
+        COMPILER_MEMORY_FENCE;
+
+        new_tile_group_header->SetTransactionId(new_version.offset,
+                                                INVALID_TXN_ID);
+        tile_group_header->SetTransactionId(tuple_slot, INITIAL_TXN_ID);
+
+      } else if (tuple_entry.second == RW_TYPE_INSERT) {
+        tile_group_header->SetEndCommitId(tuple_slot, MAX_CID);
+        tile_group_header->SetBeginCommitId(tuple_slot, MAX_CID);
+
+        COMPILER_MEMORY_FENCE;
+
+        tile_group_header->SetTransactionId(tuple_slot, INVALID_TXN_ID);
+      } else if (tuple_entry.second == RW_TYPE_INS_DEL) {
+        tile_group_header->SetEndCommitId(tuple_slot, MAX_CID);
+        tile_group_header->SetBeginCommitId(tuple_slot, MAX_CID);
+
+        COMPILER_MEMORY_FENCE;
+
+        tile_group_header->SetTransactionId(tuple_slot, INVALID_TXN_ID);
+      }
+    }
+  }
+
+  EndTransaction();
+
+  return Result::RESULT_ABORTED;
+}
+
+bool EagerWriteTxnManager::CauseDeadLock() {
+  LOG_INFO("Detecting dead lock");
+  std::lock_guard<std::mutex> lock(running_txn_map_mutex_);
+
+  // Always acquire the running_txn_map_mutex_ before acquiring the per context
+  // lock
+  std::unordered_set<txn_id_t> visited;
+  auto current_tid = current_txn->GetTransactionId();
+
+  std::queue<txn_id_t> traverse;
+
+  // init
+  for (auto tid : current_txn_ctx->wait_list_) {
+    LOG_INFO("visit %lu", tid);
+    traverse.push(tid);
+  }
+
+  // BFS start from current txn to detect ring
+  while (!traverse.empty()) {
+    auto tid = traverse.front();
+    traverse.pop();
+    if (running_txn_map_.count(tid) == 0) {
+      continue;
+    }
+    if (tid == current_tid) {
+      LOG_INFO("Find dead lock");
+      return true;
+    }
+    for (auto ttid : running_txn_map_[tid]->wait_list_) {
+      if (visited.count(ttid) != 0) {
+        continue;
+      }
+      traverse.push(ttid);
+      visited.insert(ttid);
+    }
+  }
+
+  return false;
+}
+}
+}
```

Comment:
It seems like we can't view any of the files past this point in the pull request, due to too many files being included. This means that we can't provide inline feedback for the rest of the files. Any suggestions on how to get around this? @apavlo @ranxian